### PR TITLE
Add ExpandDims layer of tf_importer.cpp

### PR DESCRIPTION
**Pull request by zihaomu**
The `ExpandDims` layer was implemented by "Reshape layer."
The following code will generate the `ExpandDims` layer:
```py
expand_dim = tf.expand_dims(input, -1)
# or
expand_dim = tf.expand_dims(input, 2)

```
Related test data could be found at [PR in opencv_extra](https://github.com/opencv/opencv_extra/pull/911).

There are still some unresolved issues here:
- 1. ~~**Currently, this code can only handle when the dimension input layer dimension is 4.**, since the different Datalayout between OpenCV and Tensorflow.~~(Solved.)
- 2. Since the implementation uses the `dstNet.getLayerShapes()` function. This implementation can not handle [this model](https://drive.google.com/drive/folders/1-y9zh3fiwARdAXjL41X-2vNwtr9Phjqi?usp=sharing). 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
opencv_extra=tf_expand_dim_layer
```

### CODE REVIEW

**Comment by zihaomu** [comment]
jenkins cn please retry a build

**Comment by fengyuentau** [comment]
jenkins cn please retry a build

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 1]
```diff
@@ -1113,6 +1115,74 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+    dstNet.getLayerShapes(netInputShapes, inpIdindex, inShape_, outShape_);
+    MatShape inpShape = outShape_[0];
+    std::vector<int> outShape = inpShape;
+
+    int outShapeSize = outShape.size();
+
+    // 2nd blob is dims tensor
+    int axis = getConstBlob(layer, value_id, 1).int_val().Get(0);
+
+    if(axis == -1)
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1140]: Quoting TF [docs](https://www.tensorflow.org/api_docs/python/tf/expand_dims), "Given an input of D dimensions, axis can be in range [-(D+1), D] (inclusive)."

Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1140]: Quoting TF [docs](https://www.tensorflow.org/api_docs/python/tf/expand_dims), "Given an input of D dimensions, axis can be in range [-(D+1), D] (inclusive)."

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 2]
```diff
@@ -1113,6 +1115,74 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+    CV_CheckGT(num_inputs, 0, "");
+    Pin inpId = parsePin(layer.input(0));
+    DataLayout inpLayout = getDataLayout(layer.input(0), data_layouts);
+
+    // Get input shape
+    std::vector<MatShape> inShape_, outShape_;
+    int inpIdindex = layer_id.find(inpId.name)->second;
+
+    dstNet.getLayerShapes(netInputShapes, inpIdindex, inShape_, outShape_);
+    MatShape inpShape = outShape_[0];
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1132]: Can you insert an assert, stating that we do not support inpShape.size() != 4?

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 3]
```diff
@@ -1113,6 +1115,74 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+            {
+                std::swap(outShape[i], outShape[i+1]);
+            }
+        }
+    }
+
+    if(inpLayout == DATA_LAYOUT_NHWC)
+    {
+        // toNCHW
+        axis = (axis != 0)?(axis % outShapeSize + 1):0;
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1165]: Can you leave the comment about how this works? And check that for all axis in [-5; 4] this works correctly and add this test to extra?

**Review by rogday — CHANGES_REQUESTED** [review]
Thank you for contribution! Can you clarify, under which conditions this implementation of ExpandDims will work? My understanding is that the network should specify shapes and they shouldn't be dynamic, is that correct? Also, could you rebase extra, so that it can be tested on CI?

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 3 above)
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1165]: Thank you for your code review, I will update your comment in the next version.

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 2 above)
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1132]: Ok, we can support inpShape.size() == 4 first.

Comment (rogday) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1132]: Your post states, that it doesn't support anything else. Is that still correct?

Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1132]: Hi @rogday, My idea is to let `ExpandDims` support different inputs size, not just 4 dimension. But here are some things I don't understand: 
1. When I use the `DataLayout inpLayout = getDataLayout(layer.input(0), data_layouts);` to get the input data Layout, I found the given `inpLayout` equals to `DATA_LAYOUT_NHWC`. But when I print the size of data, the actual data layout equals `DATA_LAYOUT_NCHW`. Does this mean that tf_importer will automatically convert `NHWC` data into `NCHW` data?

Comment (rogday) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1132]: inpLayout is unknown when there are no previous nodes with data_layput attribute(otherwise there will be data layout propagation). I think, TF defaults to NHWC. Since OpenCV layers mostly can only operate on NCHW layout, I think it converts only attributes and constants because input is expected to be nchw.

Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1134]: @rogday Thank you for your reply, I have updated the code, which could handle 5-dim and mutil-dim input data. And the test data was updated, too. You can found the new test data at [PR of opencv_extra](https://github.com/opencv/opencv_extra/pull/911).

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 1 above)
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: Thanks for the code review, it was solved at the latest code.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 4]
```diff
@@ -1113,6 +1115,94 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
     }
 }
 
+void TFImporter::parseExpandDims(tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams)
+{
+      const std::string& name = layer.name();
+    const int num_inputs = layer.input_size();
+
+    CV_Assert(!netInputShapes.empty());
```
Comment (zihaomu) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1123]: When `netInputShapes` is empty, it will not work.

Comment (rogday) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1123]: I'm not sure there is. I tried adding shape propagation to TF importer some time ago, similar to ONNX importer, but realized that netInputShapes aren't required to exist.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 5]
```diff
@@ -2813,13 +2912,7 @@ void TFImporter::populateNet()
...
                 if (layout != DATA_LAYOUT_UNKNOWN)
                 {
-                    if (it->second == DATA_LAYOUT_UNKNOWN)
-                        it->second = layout;
-                    else if (it->second != layout)
-                    {
-                        it->second = DATA_LAYOUT_UNKNOWN;
-                        layout = DATA_LAYOUT_UNKNOWN;
-                    }
+                    it->second = layout;
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:2915]: Is this change necessary? This doesn't seem to affect tests. If this fixes something in your model, I propose to open a new PR, describing the problem.

**Review by rogday — CHANGES_REQUESTED** [review]
Looks good with a few comments. :+1:

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 6]
```diff
@@ -1113,6 +1115,94 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+    std::vector<MatShape> inShape_, outShape_;
+    int inpIdindex = layer_id.find(inpId.name)->second;
+
+    dstNet.getLayerShapes(netInputShapes, inpIdindex, inShape_, outShape_);
+    MatShape inpShape = outShape_[0];
+    std::vector<int> outShape = inpShape;
+
+    int outShapeSize = outShape.size();
+
+    CV_Assert(inpShape.size() >= 1);
```
[the remark this answers is not in the document]
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1139]: Thanks, I will check it out.

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 5 above)
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:2915]: I will create a new PR for that.

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1139]: @rogday Hi, the 3-dim input was supported in the last code.

[the remark this answers is not in the document]
Comment (rogday) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1139]: @zihaomu, I ran this code to generate models:

```python

for d in range(1, 7):
    for axis in range(-d - 1, d + 1):
        inp = tf.placeholder(tf.float32, list(range(1, d + 1)), 'input')
        expand_dim = inp + 1
        expand_dim = tf.expand_dims(expand_dim, axis)
        save(inp, expand_dim, f'expand_dims_{d}_{axis}', optimize=False)

```

Note, that I needed to modify writeBlob in the following way to ensure C_CONTIGUOUS:

```python

def writeBlob(data, name):
    if data.ndim == 4:
        # NHWC->NCHW
        data = data.transpose(0, 3, 1, 2)
    elif data.ndim == 5:
        # NDHWC->NCDHW
        data = data.transpose(0, 4, 1, 2, 3)
    np.save(name + '.npy', np.ascontiguousarray(data, np.float32))

```

and tested it on c++ side:

```cpp

    for (int d = 1; d < 7; ++d)
    {
        for (int axis = -d - 1; axis != d + 1; ++axis)
        {
            std::cout << "Running: dims=" << d << ", axis=" << axis << std::endl;
            try{
                runTensorFlowNet(cv::format("expand_dims_%d_%d", d, axis));
                std::cout << "Ok." << std::endl;
            }
            catch (const std::exception& e)
            {
                std::cout << "Failed: " << e.what() << std::endl;
            }
        }
    }

```

This test failed on dims=3, axis=0 and dims=3, axis=-4. Could you check this out?

[the remark this answers is not in the document]
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp:1139]: Thanks for code reviewing, I have updated the code.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 7]
```diff
@@ -1113,6 +1115,120 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+    // 2nd blob is dims tensor
+    int axis = getConstBlob(layer, value_id, 1).int_val().Get(0);
+
+    // Convert negative numbers to positive numbers, axis can be in range [-(D+1), D].
+    if(axis < 0)
+    {
+        axis = inpShape.size() + axis + 1;
+    }
+
+    CV_Assert( 0 <= axis && axis <= inpShape.size());
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1149]: Please, remove space between  ( and 0

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 8]
```diff
@@ -1113,6 +1115,120 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+            std::swap(outShape[2], outShape[3]);
+            std::swap(outShape[3], outShape[4]);
+        }
+    }
+
+    outShape.insert(outShape.begin() + axis, 1);
+    outShapeSize += 1;
+
+    // From OpenCV's NCDHW to NDHWC.
+    if((inpLayout != DATA_LAYOUT_NHWC && inpLayout != DATA_LAYOUT_NCHW) && outShapeSize == 5 )
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1207]: Please, remove space between 5 and ).

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 9]
```diff
@@ -1113,6 +1115,120 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+
+    if(outShapeSize == 5)
+    {
+        data_layouts[name] = DATA_LAYOUT_NDHWC;
+    }else if(outShapeSize == 4)
+    {
+        data_layouts[name] = DATA_LAYOUT_NCHW;
+    }else{
+        data_layouts[name] = inpLayout;
+    }
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1229]: Please, change this to

```cpp

    if (outShapeSize == 5)
    {
        data_layouts[name] = DATA_LAYOUT_NDHWC;
    }
    else if(outShapeSize == 4)
    {
        data_layouts[name] = DATA_LAYOUT_NCHW;
    }
    else
    {
        data_layouts[name] = inpLayout;
    }

```

To be consistent with style in the rest of the file.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 10]
```diff
@@ -1113,6 +1115,120 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
...
+
+    // After ExpendDims, 3-dim data will become 4-dim data, and OpenCV retains 4-dim data as NCHW data layout.
+    // Convert OpenCV's NHC to NCH first.
+    if(outShapeSize == 3)
+    {
+        // If axis equal to outShapeSize, that mean we expand in Channel dimmension, and do not add permuteLayer.
+        if(axis != outShapeSize)
+        {
+            int order[] = {0, 2, 1};  // From OpenCV's NHC to NCH.
+            addPermuteLayer(order, name + "/nhc", inpId, 3);
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:1159]: I think, you meant /nch

**Review by rogday — APPROVED** [review]
Please, add a space between ifs and fors and opening parentheses. Otherwise looks good, thank you! :+1:

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 10 above)
Comment (zihaomu) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: Thank you, all your mentions have been fixed.
